include/wil/network.h

```diff
@@ -0,0 +1,1372 @@
+        return ::wil::network::resolve_name(localhost_address_name_string);
+    }
+#endif
+
+    struct WINSOCK_EXTENSION_FUNCTION_TABLE
+    {
+        LPFN_ACCEPTEX AcceptEx{nullptr};
+        LPFN_CONNECTEX ConnectEx{nullptr};
+        LPFN_DISCONNECTEX DisconnectEx{nullptr};
+        LPFN_GETACCEPTEXSOCKADDRS GetAcceptExSockaddrs{nullptr};
+        LPFN_TRANSMITFILE TransmitFile{nullptr};
+        LPFN_TRANSMITPACKETS TransmitPackets{nullptr};
+        LPFN_WSARECVMSG WSARecvMsg{nullptr};
+        LPFN_WSASENDMSG WSASendMsg{nullptr};
+    };
+
+    struct WINSOCK_SOCKET_NOTIFICATION_TABLE
+    {
+        using LPFN_PROCESS_SOCKET_NOTIFICATION = DWORD(WSAAPI*)(
+            HANDLE completionPort,
+            UINT32 registrationCount,
+            _Inout_updates_opt_(registrationCount) SOCK_NOTIFY_REGISTRATION* registrationInfos,
+            UINT32 timeoutMs,
+            ULONG completionCount,
+            _Out_writes_to_opt_(completionCount, *receivedEntryCount) OVERLAPPED_ENTRY* completionPortEntries,
+            _Out_opt_ UINT32* receivedEntryCount);
+
+        LPFN_PROCESS_SOCKET_NOTIFICATION ProcessSocketNotifications{nullptr};
+    };
+
+    template <typename F>
+    struct socket_extension_function_table_t;
+    using winsock_extension_function_table = socket_extension_function_table_t<WINSOCK_EXTENSION_FUNCTION_TABLE>;
+    using rio_extension_function_table = socket_extension_function_table_t<RIO_EXTENSION_FUNCTION_TABLE>;
+    using process_socket_notification_table = socket_extension_function_table_t<WINSOCK_SOCKET_NOTIFICATION_TABLE>;
+
+    template <typename F>
+    struct socket_extension_function_table_t
+    {
+        socket_extension_function_table_t() WI_NOEXCEPT : wsa_reference_count{WSAStartup_nothrow()}
+        {
+            load();
+        }
+
+        // can copy, but the new object needs its own WSA reference count
+        // (getting a WSA reference count should be no-fail once the first reference it taken)
+        //
+        // IF the target could not get a WSA reference count
+        //   OR
+        // IF we couldn't get our own WSA reference count
+        //   (this should never happen the caller has a reference, but we failed to get a WSA reference)
+        // THEN
+        //   this object cannot carry forward any function pointers - it must show successfully loaded == false
+        socket_extension_function_table_t(const socket_extension_function_table_t& rhs) WI_NOEXCEPT
+            : wsa_reference_count{WSAStartup_nothrow()}
+        {
+            if (!wsa_reference_count || !rhs.wsa_reference_count)
+            {
+                ::memset(&function_table, 0, sizeof(function_table));
+            }
+            else
+            {
+                ::memcpy_s(&function_table, sizeof(function_table), &rhs.function_table, sizeof(rhs.function_table));
+            }
+        }
+
+        socket_extension_function_table_t& operator=(const socket_extension_function_table_t& rhs) WI_NOEXCEPT

```

Comment:
What is the purpose to having assignment operators on this type? You declare the reference count as const, neutering your ability to take ownership, meaning assignment will only ever work if _both_ are properly initialized, but at that point AFAICT you're just overwriting the function table with the same values.

Reply:
Great points. I don't think assignment operators are actually needed.
Will look more into this.